syntax = "proto3";

package raft;

option cc_generic_services = true;

message RequestVoteRequest
{
    uint64 term = 1;
    uint32 candidateId = 2;
    uint64 lastLogIndex = 3;
    uint64 lastLogTerm = 4;
}

message RequestVoteResponse
{
    uint64 term = 1;
    bool voteGranted = 2;
}

message LogEntry
{
    uint64 term = 1;
    bytes command = 2;
}

message AppendEntriesRequest
{
    uint64 term = 1;
    uint32 leaderId = 2;
    uint64 prevLogIndex = 3;
    uint64 prevLogTerm = 4;
    repeated LogEntry entries = 5;
    uint64 leaderCommit = 6;
}

message AppendEntriesResponse
{
    uint64 term = 1;
    bool success = 2;
}

message InstallSnapshotRequest
{
    uint64 term = 1;
    uint32 leaderId = 2;
    uint64 lastIncludedIndex = 3;
    uint64 lastIncludedTerm = 4;
    bytes data = 5;
}

message InstallSnapshotResponse
{
    uint64 term = 1;
}

service RaftService
{
    rpc RequestVoteRPC(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntriesRPC(AppendEntriesRequest) returns (AppendEntriesResponse);
    rpc InstallSnapshotRPC(InstallSnapshotRequest) returns (InstallSnapshotResponse);
}